### ISSUE TITLE
Allow specifying a default image when imagemagick is not found

### ISSUE THREAD

**Issue by nixjdm** [lektor/lektor#799]
From https://github.com/lektor/lektor/issues/363#issuecomment-661217233:

> An option would be to have .thumbnail() issue a console error/warning message, but return a jinja2.Undefined instance. This would allow for explicit fallback handling in the template, should the user want. E.g.
>
> ```
> <% set thumb = image.thumbnail()|default(image) %>
> <img src="{{ thumb|url }}">
> ```
>
> (I'm not positive, but from brief inspection of the current code, it appears that currently in this case, .thumbnail() returns a valid Thumbnail instance, so there is no sign of failure available by inspecting the return value. The failure isn't reported until later when the attempt is made to build the thumbnail artifact.)

With the appropriately placed `return jinja2.Undefined("Could not locate Imagemagic")`, we can allow people to catch this particular error and set a default if they wish. If they don't, a useful traceback would still be provided, and build errors encountered.

I can see a use case for this where a developer specifies an single image that is an obvious stand-in, so a content producer could still effectively use the site locally, knowing what they see is an approximate facsimile, so they can just add content. ~~As an extension, to make the dummy images play well with all pages, a developer could again have `.thumbnail()` calls on the default image to rescale them correctly.~~

**Comment by xlotlu** [comment]
> .thumbnail() returns a valid Thumbnail instance, so there is no sign of failure available by inspecting the return value. The failure isn't reported until later when the attempt is made to build the thumbnail artifact.)

This is expected behaviour, unless you want to blow up the build process right there and then. The point is the Thumbnail instance has width, height and other attributes which are used in the template. Returning anything else will blow up elsewhere instead.

@dairiki and @nixjdm, to me this feels like bike-shedding a wrongly identified issue. Think about it this way: would you jump through all these hoops if there were no imagemagick involved, and the code was pure Python, and the thumbnail generation failed because of <whatever>, say, a missing dependency?

If the answer is no, then this is the wrong solution to the wrong problem. To me the problem is that lektor can exist without imagemagick installed. Which is only fine if the user never uses thumbnails, otherwise we should either:

a) point the user to installing it themselves, and making sure it *always* works (which I'm pretty sure it does at this point), or
b) not ship the product without this very essential dependency. E.g. make sure that `chocolatey install lektor` pulls imagemagick.

**Comment by nixjdm** [comment]
Yes, actually I think you're right. I rushed a bit and didn't think it through right. I do think it's a rather essential tool, and (after realizing my silly strike-throughed mistake above) I struggle to think of when you'd really want to build a template without the image resizing you expect. I guess you could resize in other ways, like CSS or JS, but you'd be doing that for every thumbnail operation to support that. Maybe someone would want that, but ...

we could just download it. For example, I think ideally we would minimally pollute the users system, so not broadly install with a chocolately/brew/yum/etc. Looking at https://imagemagick.org/script/download.php, it looks like imagemagick has portable binaries for most systems. We could have the setup.py download it and use that. That would basically make this behave like all of the other dependencies we have, installable and uninstallable with pip. I've actually done that before.

ffmpeg has the same options I think.

**Comment by xlotlu** [comment]
I meant `choco install` would pull imagemagick (and ffmpeg) as dependencies. True package management. And yes, I guess we should have two installers, one with dependencies bundled  and the other without.

So I think we should focus our efforts towards such an installer. I just figured out python is a "dependency" itself.

> I struggle to think of when you'd really want to build a template without the image resizing you expect.

In the spirit of the original idea, the code should return a Thumbnail instance, but with the path set to some empty / transparent "error image". This would cater to the occasional individual that: a) inherits a website from another developer; b) doesn't manage to install imagemagick (eyes rolling, they just managed to install lektor); c) needs to edit content; d) really doesn't care about images.

I really think this is energy better spent elsewhere. :)
